probe/sniff/sniffer.go

```diff
@@ -0,0 +1,247 @@
+package sniff
+
+import (
+	"io"
+	"log"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/weaveworks/scope/report"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+// Sniffer is a packet-sniffing reporter.
+type Sniffer struct {
+	hostID  string
+	reports chan chan report.Report
+	parser  *gopacket.DecodingLayerParser
+	decoded []gopacket.LayerType
+	eth     layers.Ethernet
+	ip4     layers.IPv4
+	ip6     layers.IPv6
+	tcp     layers.TCP
+	udp     layers.UDP
+	icmp4   layers.ICMPv4
+	icmp6   layers.ICMPv6
+}
+
+// New returns a new sniffing reporter that samples traffic by turning its
+// packet capture facilities on and off. Note that the on and off durations
+// represent a way to bound CPU burn. Effective sample rate needs to be
+// calculated as (packets decoded / packets observed).
+func New(hostID string, src gopacket.ZeroCopyPacketDataSource, on, off time.Duration) *Sniffer {
+	s := &Sniffer{
+		hostID:  hostID,
+		reports: make(chan chan report.Report),
+	}
+	s.parser = gopacket.NewDecodingLayerParser(
+		layers.LayerTypeEthernet,
+		&s.eth, &s.ip4, &s.ip6, &s.tcp, &s.udp, &s.icmp4, &s.icmp6,
+	)
+	go s.loop(src, on, off)
+	return s
+}
+
+// Report implements the Reporter interface.
+func (s *Sniffer) Report() (report.Report, error) {
+	c := make(chan report.Report)
+	s.reports <- c
+	return <-c, nil
+}
+
+func (s *Sniffer) loop(src gopacket.ZeroCopyPacketDataSource, on, off time.Duration) {
+	var (
+		process = uint64(1)               // initially enabled
+		total   = uint64(0)               // total packets seen
+		count   = uint64(0)               // count of packets captured
+		packets = make(chan Packet)       // decoded packets
+		rpt     = report.MakeReport()     // the report we build
+		turnOn  = (<-chan time.Time)(nil) // signal to start capture (initially enabled)
+		turnOff = time.After(on)          // signal to stop capture
+		done    = make(chan struct{})     // when src is finished, we're done too
+	)
+
+	go func() {
+		s.read(src, packets, &process, &total, &count)
+		close(done)
+	}()
+
+	for {
+		select {
+		case p := <-packets:
+			s.Merge(p, rpt)
```

Comment:
In the on state, yes. Good catch. I think it'd be better to do a nonblocking send to the packets chan, what do you think?